### PR TITLE
Support for Channel Upgrade

### PR DESCRIPTION
**Pull request by siburu**

### CODE REVIEW

File: pkg/relay/ethereum/chain.go [hunk 1]
```diff
@@ -474,6 +474,42 @@ func (c *Chain) QueryUnfinalizedRelayAcknowledgements(ctx core.QueryContext, cou
...
+		return nil, err
+	} else if ev == nil {
+		return nil, nil
+	} else {
+		return &chantypes.QueryUpgradeErrorResponse{
+			ErrorReceipt: chantypes.ErrorReceipt{
+				Sequence: ev.UpgradeSequence,
+				Message:  ev.Message,
+			},
+			ProofHeight: clienttypes.NewHeight(0, ev.Raw.BlockNumber),
```
Comment (bluele) [inline pkg/relay/ethereum/chain.go:501]: Do we need to set this proof height?

File: pkg/relay/ethereum/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		timeout,
+	)
+	if err != nil {
+		return err
+	}
+
+	if receipt, err := c.client.WaitForReceiptAndGet(ctx, tx.Hash()); err != nil {
+		return err
+	} else if receipt.Status == gethtypes.ReceiptStatusFailed {
+		return fmt.Errorf("tx execution reverted")
```
Comment (bluele) [inline pkg/relay/ethereum/upgrade.go:50]: We should try to retrieve a revert reason from receipt here

File: pkg/relay/ethereum/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		upgradeSequence,
+	)
+	if err != nil {
+		return err
+	}
+
+	if receipt, err := c.client.WaitForReceiptAndGet(ctx, tx.Hash()); err != nil {
+		return err
+	} else if receipt.Status == gethtypes.ReceiptStatusFailed {
+		return fmt.Errorf("tx execution reverted")
```
Comment (bluele) [inline pkg/relay/ethereum/upgrade.go:93]: same as https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/50/files#r1730883729

File: pkg/relay/ethereum/chain.go (hunk 1 above)
Comment (siburu) [reply pkg/relay/ethereum/chain.go:501]: No! Thank you I will delete this line.

File: pkg/relay/ethereum/upgrade.go (hunk 2 above)
Comment (siburu) [reply pkg/relay/ethereum/upgrade.go:50]: Sure. I will try to do that.

File: pkg/relay/ethereum/chain.go (hunk 1 above)
Comment (siburu) [reply pkg/relay/ethereum/chain.go]: @bluele done: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/50/commits/91f370e31062b4e6f446dc059a412eb96b272da8

File: pkg/relay/ethereum/upgrade.go (hunk 2 above)
Comment (siburu) [reply pkg/relay/ethereum/upgrade.go]: @bluele done: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/50/commits/2c0371da76e0fc01df339f5c78c0d79100b116d7

File: pkg/relay/ethereum/upgrade.go (hunk 3 above)
Comment (siburu) [reply pkg/relay/ethereum/upgrade.go:188]: @bluele done: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/50/commits/2c0371da76e0fc01df339f5c78c0d79100b116d7

File: contracts/IBCContractUpgradableModule.sol [hunk 4]
```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.12;
```
Comment (bluele) [inline contracts/IBCContractUpgradableModule.sol:2]: want to align it with the minimal version of ibc-solidity, which is v0.8.20

Comment (siburu) [reply contracts/IBCContractUpgradableModule.sol:2]: OK. The following files haven't been updated yet!
https://github.com/hyperledger-labs/yui-ibc-solidity/blob/v0.3.35/contracts/core/04-channel/IBCChannelUpgrade.sol#L2
https://github.com/hyperledger-labs/yui-ibc-solidity/blob/v0.3.35/contracts/core/26-router/IIBCModuleUpgrade.sol#L2

Comment (bluele) [reply contracts/IBCContractUpgradableModule.sol]: Oh, I missed those. Thanks!

**Review by bluele — APPROVED** [review]
@siburu LGTM👍

**Pull request merged**